std/digest/hmac.d

```diff
@@ -0,0 +1,305 @@
+            buffer = digest.finish();
+            secret = buffer[];
+        }
+
+        // if secret is too short, it will be padded with zeroes
+        // (the key buffer is already zero-initialized)
+        import std.algorithm.mutation : copy;
+        secret.copy(key[]);
+
+        start();
+    }
+
+    ///
+    @safe pure nothrow @nogc unittest
+    {
+        import std.digest.sha, std.digest.hmac;
+        import std.string : representation;
+        auto hmac = HMAC!SHA1("My s3cR3T keY".representation);
+    }
+
+    /**
+     * Reinitializes the digest, making it ready for reuse.
+     *
+     * Note:
+     * The constructor leaves the digest in an initialized state, so that this
+     * method only needs to be called if an unfinished digest is to be reused.
+     *
+     * Returns:
+     * A reference to the digest for convenient chaining.
+     */
+
+    ref start() return
+    {
+        ubyte[blockSize / 8] ipad = void;
+        foreach(immutable i; 0 .. blockSize / 8)
+            ipad[i] = key[i] ^ 0x36;
+
+        digest.start();
+        digest.put(ipad[]);
+
+        return this;
+    }
+
+    ///
+    @safe pure nothrow @nogc unittest
+    {
+        import std.digest.sha, std.digest.hmac;
+        import std.string : representation;
+        ubyte[] data1, data2;
+        auto hmac = HMAC!SHA1("My s3cR3T keY".representation);
+        hmac.put(data1);
+        hmac.start();    // reset digest
+        hmac.put(data2); // start over
+    }
+
+    /**
+     * Feeds a piece of data into the hash computation. This method allows the
+     * type to be used as an $(XREF range, OutputRange).
+     *
+     * Returns:
+     * A reference to the digest for convenient chaining.
+     */
+
+    ref put(in ubyte[] data...) return
+    {
+        digest.put(data);
+        return this;
+    }
+
+    ///
+    @safe pure nothrow @nogc unittest
+    {
+        import std.digest.sha, std.digest.hmac;
+        import std.string : representation;
+        ubyte[] data1, data2;
+        auto hmac = HMAC!SHA1("My s3cR3T keY".representation);
+        hmac.put(data1)
+            .put(data2);
+    }
+
+    /**
+     * Resets the digest and returns the finished hash.
+     */
+
+    DigestType!H finish()
+    {
+        ubyte[blockSize / 8] opad = void;
+        foreach(immutable i; 0 .. blockSize / 8)
+            opad[i] = key[i] ^ 0x5c;
+
+        auto tmp = digest.finish();
+
+        digest.start();
+        digest.put(opad[]);
+        digest.put(tmp);
+        auto result = digest.finish();
+        start();    // reset the digest
+        return result;
+    }
+
+    ///
+    @safe pure nothrow @nogc unittest
+    {
+        import std.digest.sha, std.digest.hmac;
+        import std.string : representation;
+        ubyte[] data1, data2;
+        auto hmac = HMAC!SHA1("My s3cR3T keY".representation);
+        auto digest = hmac.put(data1)
+                          .put(data2)
+                          .finish();
+    }
```

Comment:
verify that the digest is as expected (again data1 and data2 should not be empty)